File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/AccountIntegrationsLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.navigation.MobilePlatformFeatures
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import java.net.URLEncoder

/** Account integrations helpers (M14.1). */
object AccountIntegrationsLogic {
    const val RBAC_MANAGE_PERMISSION = "global:app:rbac:manage"
    val defaultCreateScopes = listOf("mcp:connect", "courses:read")

    fun integrationsEnabled(features: MobilePlatformFeatures): Boolean =
        features.ffMobileSettingsIntegrations

    fun accessKeysEnabled(features: MobilePlatformFeatures): Boolean =
        features.ffApiTokens

    fun calendarSubscriptionsEnabled(features: MobilePlatformFeatures): Boolean =
        features.ffCalendarFeeds

    fun canManageServiceTokens(permissions: List<String>): Boolean =
        permissions.contains(RBAC_MANAGE_PERMISSION)

    fun shouldShowServiceTokensSection(
        permissions: List<String>,
        adminApiForbidden: Boolean,
    ): Boolean = !adminApiForbidden && canManageServiceTokens(permissions)

    fun isRevoked(revokedAt: String?): Boolean = !revokedAt.isNullOrEmpty()

    fun activeAccessKeys(tokens: List<AccessKeySummary>): List<AccessKeySummary> =
        tokens.filter { !isRevoked(it.revokedAt) && it.isServiceToken != true }

    fun revokedAccessKeys(tokens: List<AccessKeySummary>): List<AccessKeySummary> =
        tokens.filter { isRevoked(it.revokedAt) && it.isServiceToken != true }

    fun activeServiceTokens(tokens: List<AccessKeySummary>): List<AccessKeySummary> =
        tokens.filter { it.isServiceToken == true && !isRevoked(it.revokedAt) }

    fun resolveCalendarFeedURL(template: String, token: String): String {
        val encoded = URLEncoder.encode(token, "UTF-8")
        return template.replace("<token>", encoded)
    }

    fun resolvedPersonalFeedUrl(
        info: CalendarTokenInfo?,
        created: CalendarTokenCreated?,
    ): String? {
        created?.feedUrl?.takeIf { it.isNotBlank() }?.let { return it }
        val token = created?.token ?: return null
        val template = info?.personalFeedUrl ?: return null
        if (template.isBlank()) return null
        return resolveCalendarFeedURL(template, token)
    }

    fun resolvedCourseFeedUrl(template: String, token: String?): String? {
        if (token.isNullOrBlank()) return null
        return resolveCalendarFeedURL(template, token)
    }

    fun mcpConfigJson(base: JsonObject, token: String): String {
        val trimmed = token.trim()
        if (trimmed.isEmpty()) return base.toString()
        val servers = base["mcpServers"]?.jsonObject ?: return base.toString()
        val lextures = servers["lextures"]?.jsonObject ?: return base.toString()
        val env = lextures["env"]?.jsonObject ?: return base.toString()
        val updatedEnv = buildJsonObject {
            env.forEach { (key, value) -> put(key, value) }
            put("LEXTURES_API_TOKEN", JsonPrimitive(trimmed))
        }
        val updatedLextures = buildJsonObject {
            lextures.forEach { (key, value) ->
                put(key, if (key == "env") updatedEnv else value)
            }
        }
        val updatedServers = buildJsonObject {
            servers.forEach { (key, value) ->
                put(key, if (key == "lextures") updatedLextures else value)
            }
        }
        return buildJsonObject {
            base.forEach { (key, value) ->
                put(key, if (key == "mcpServers") updatedServers else value)
            }
        }.toString()
    }
}

File: clients/ios/Lextures/Core/LMS/AccountIntegrationsLogic.swift
import Foundation

/// Account integrations helpers (M14.1) — gating, secret handling, and feed URL resolution.
enum AccountIntegrationsLogic {
    static let rbacManagePermission = "global:app:rbac:manage"
    static let defaultCreateScopes = ["mcp:connect", "courses:read"]

    static func integrationsEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffMobileSettingsIntegrations
    }

    static func accessKeysEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffApiTokens
    }

    static func calendarSubscriptionsEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffCalendarFeeds
    }

    static func canManageServiceTokens(permissions: [String]) -> Bool {
        permissions.contains(rbacManagePermission)
    }

    /// Whether the service-tokens section may be shown (permission gate; org admins resolved via API probe).
    static func shouldShowServiceTokensSection(
        permissions: [String],
        adminApiForbidden: Bool
    ) -> Bool {
        !adminApiForbidden && canManageServiceTokens(permissions: permissions)
    }

    static func isRevoked(_ revokedAt: String?) -> Bool {
        guard let revokedAt else { return false }
        return !revokedAt.isEmpty
    }

    static func activeAccessKeys(_ tokens: [AccessKeySummary]) -> [AccessKeySummary] {
        tokens.filter { !isRevoked($0.revokedAt) && $0.isServiceToken != true }
    }

    static func revokedAccessKeys(_ tokens: [AccessKeySummary]) -> [AccessKeySummary] {
        tokens.filter { isRevoked($0.revokedAt) && $0.isServiceToken != true }
    }

    static func activeServiceTokens(_ tokens: [AccessKeySummary]) -> [AccessKeySummary] {
        tokens.filter { $0.isServiceToken == true && !isRevoked($0.revokedAt) }
    }

    static func resolveCalendarFeedURL(template: String, token: String) -> String {
        let encoded = token.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? token
        return template.replacingOccurrences(of: "<token>", with: encoded)
    }

    static func resolvedPersonalFeedURL(
        info: CalendarTokenInfo?,
        created: CalendarTokenCreated?
    ) -> String? {
        if let feedUrl = created?.feedUrl, !feedUrl.isEmpty { return feedUrl }
        if let token = created?.token, let template = info?.personalFeedUrl, !template.isEmpty {
            return resolveCalendarFeedURL(template: template, token: token)
        }
        return nil
    }

    static func resolvedCourseFeedURL(template: String, token: String?) -> String? {
        guard let token, !token.isEmpty else { return nil }
        return resolveCalendarFeedURL(template: template, token: token)
    }

    /// Injects an access key into MCP JSON config for copy-to-clipboard.
    static func mcpConfigJSON(base: [String: JSONValue], token: String) -> String? {
        var draft = base
        guard case .object(var servers)? = draft["mcpServers"],
              case .object(var lextures)? = servers["lextures"],
              case .object(var env)? = lextures["env"] else {
            return prettyJSONString(base)
        }
        let trimmed = token.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmed.isEmpty {
            env["LEXTURES_API_TOKEN"] = .string(trimmed)
        }
        lextures["env"] = .object(env)
        servers["lextures"] = .object(lextures)
        draft["mcpServers"] = .object(servers)
        return prettyJSONString(draft)
    }

    private static func prettyJSONString(_ value: [String: JSONValue]) -> String? {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let data = try? encoder.encode(value) else { return nil }
        return String(data: data, encoding: .utf8)
    }
}